### PR TITLE
Add DTO to prevent over-posting on API

### LINKED ISSUE #15204: Provide examples for DTOs to prevent over-posting
Using DTOs should be described in this article. I know there is a small part about overposting in "Data Access", but especially for APIs there should be a concrete example of using DTOs and mapping them, with for example AutoMapper, to domain objects. It also enforces security by thinking about what properties you want to serialize to the client. This should be mentioned as best practice. In the old Web API tutorials the term DTOs was used, now only the term ViewModel is used, because its used with Razor Pages in mind.

This was in the old docs: https://docs.microsoft.com/en-us/aspnet/web-api/overview/data/using-web-api-with-entity-framework/part-5

@Rick-Anderson  edit:
When this is fixed, PR scaffolding so Web API scaffolding points to this doc and not the RP/MVC approach. See sample PR at https://github.com/aspnet/Scaffolding/pull/1241

---
#### Document Details

⚠ *Do not edit this section. It is required for docs.microsoft.com ➟ GitHub issue linking.*

* ID: 583ae3e7-5915-f22b-e54c-6c100cf2a7da
* Version Independent ID: 95f2549e-29f9-e31a-b4d0-432b971643ff
* Content: [Tutorial: Create a web API with ASP.NET Core](https://docs.microsoft.com/en-us/aspnet/core/tutorials/first-web-api?view=aspnetcore-3.0&tabs=visual-studio#feedback)
* Content Source: [aspnetcore/tutorials/first-web-api.md](https://github.com/aspnet/AspNetCore.Docs/blob/master/aspnetcore/tutorials/first-web-api.md)
* Product: **aspnet-core**
* Technology: **aspnetcore-tutorials**
* GitHub Login: @Rick-Anderson
* Microsoft Alias: **riande**

### PR DESCRIPTION
**Pull request by Rick-Anderson**
Fixes #15204

[Internal review URL](https://review.docs.microsoft.com/en-us/aspnet/core/tutorials/first-web-api?view=aspnetcore-3.1&branch=pr-en-us-17074&tabs=visual-studio#prevent-over-posting)

cc @serpent5

### CODE REVIEW

File: aspnetcore/tutorials/first-web-api.md [hunk 1]
```diff
@@ -453,6 +453,37 @@ Use Postman to delete a to-do item:
 * Set the URI of the object to delete (for example `https://localhost:5001/api/TodoItems/1`).
 * Select **Send**.
 
+<a name="over-post"></a>
+
+## Prevent over-posting
+
+Currently the app exposes the entire `TodoItem`. Productions apps typically limit the data that's input and returned using subset of the model. The subset model is called a Data Transfer Object (DTO), input model, or view model.
```
Comment (mkArtakMSFT) [inline aspnetcore/tutorials/first-web-api.md]: What about trying to rephrase this first paragraph as follows? Feel free to dismiss this suggestion, if you don't like it.
```suggestion
Currently our sample app exposes the entire `TodoItem` object. Productions apps typically limit the data that's input and returned using subset of the model. There are multiple reasons behind this and security aspect is a major one. That subset of a model is usually referred to as a Data Transfer Object (DTO), input model, or view model. We will use the `DTO` naming in this article.
```

File: aspnetcore/tutorials/first-web-api.md [hunk 2]
```diff
@@ -453,6 +453,37 @@ Use Postman to delete a to-do item:
...
+
+A DTO may be used to:
+
+* Prevent over-posting.
+* Hide properties that clients are not supposed to view.
+* Omit some properties in order to reduce payload size.
+* Flatten object graphs that contain nested objects. Flattened object graphs can be more convenient for clients.
+
+Update the `TodoItem` class to include a secret field:
```
Comment (mkArtakMSFT) [inline aspnetcore/tutorials/first-web-api.md]:
```suggestion
To demonstrate this here, let's update the `TodoItem` class to include a secret field:
```

File: aspnetcore/tutorials/first-web-api.md [hunk 3]
```diff
@@ -453,6 +453,37 @@ Use Postman to delete a to-do item:
 * Set the URI of the object to delete (for example `https://localhost:5001/api/TodoItems/1`).
 * Select **Send**.
 
+<a name="over-post"></a>
+
+## Prevent over-posting
+
+Currently the sample app exposes the entire `TodoItem` object. Productions apps typically limit the data that's input and returned using a subset of the model. There are multiple reasons behind this and security aspect is a major one. The subset of a model is usually referred to as a Data Transfer Object (DTO), input model, or view model. **DTO** is used in this article.
```
Comment (serpent5) [inline aspnetcore/tutorials/first-web-api.md]:
```suggestion
Currently the sample app exposes the entire `TodoItem` object. Productions apps typically limit the data that's input and returned using a subset of the model. There are multiple reasons behind this and security is a major one. The subset of a model is usually referred to as a Data Transfer Object (DTO), input model, or view model. **DTO** is used in this article.
```

File: aspnetcore/tutorials/first-web-api.md [hunk 4]
```diff
@@ -453,6 +453,37 @@ Use Postman to delete a to-do item:
...
+* Prevent over-posting.
+* Hide properties that clients are not supposed to view.
+* Omit some properties in order to reduce payload size.
+* Flatten object graphs that contain nested objects. Flattened object graphs can be more convenient for clients.
+
+To demonstrate the DTO approach, update the `TodoItem` class to include a secret field:
+
+[!code-csharp[](first-web-api/samples/3.0/TodoApiDTO/Models/TodoItem.cs?name=snippet&highlight=6)]
+
+The secret field needs to be hidden from this app, an administrative app could expose it.
```
Comment (serpent5) [inline aspnetcore/tutorials/first-web-api.md]:
```suggestion
The secret field needs to be hidden from this app, but an administrative app could choose to expose it.
```

File: aspnetcore/tutorials/first-web-api.md [hunk 5]
```diff
@@ -453,6 +453,37 @@ Use Postman to delete a to-do item:
...
+
+Verify you can post and get the secret field.
+
+Create an DTO model:
```
Comment (serpent5) [inline aspnetcore/tutorials/first-web-api.md]:
```suggestion
Create a DTO model:
```

File: aspnetcore/tutorials/first-web-api.md [hunk 6]
```diff
@@ -453,6 +453,37 @@ Use Postman to delete a to-do item:
...
+
+[!code-csharp[](first-web-api/samples/3.0/TodoApiDTO/Models/TodoItemDTO.cs?name=snippet)]
+
+Update the `TodoItemsController` to use `TodoItemDTO`:
+
+[!code-csharp[](first-web-api/samples/3.0/TodoApiDTO/Controllers/TodoItemsController.cs?name=snippet)]
+
+Verify you can't post and get the secret field.
```
Comment (serpent5) [inline aspnetcore/tutorials/first-web-api.md]:
```suggestion
Verify you can't post or get the secret field.
```

**Pull request merged**